File: node/src/hash.ts
import { createHash } from "node:crypto";
import { readdir, readFile, stat } from "node:fs/promises";
import path from "node:path";

const IGNORE = new Set([".git", "__pycache__", ".DS_Store"]);

async function walkFiles(root: string): Promise<string[]> {
  const entries = await readdir(root, { withFileTypes: true });
  const files: string[] = [];
  for (const entry of entries) {
    if (IGNORE.has(entry.name)) continue;
    const full = path.join(root, entry.name);
    if (entry.isDirectory()) {
      files.push(...(await walkFiles(full)));
    } else if (entry.isFile()) {
      files.push(full);
    }
  }
  return files.sort();
}

export function hashFromMap(mapping: Record<string, string>): string {
  const payload = JSON.stringify(Object.fromEntries(Object.entries(mapping).sort()));
  return `sha256:${createHash("sha256").update(payload).digest("hex")}`;
}

export async function dirHash(root: string): Promise<string> {
  if (!(await stat(root).catch(() => null))) return hashFromMap({});
  const files = await walkFiles(root);
  const mapping: Record<string, string> = {};
  for (const file of files) {
    const rel = path.relative(root, file).split(path.sep).join("/");
    const bytes = await readFile(file);
    mapping[rel] = createHash("sha256").update(bytes).digest("hex");
  }
  return hashFromMap(mapping);
}

export function remoteFingerprint(blobMap: Record<string, string>): string {
  return hashFromMap(blobMap);
}

File: python/src/skills_utils/hash.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path

IGNORE_PARTS = {".git", "__pycache__", ".DS_Store"}


def _should_skip(path: Path, root: Path) -> bool:
    rel = path.relative_to(root)
    return any(part in IGNORE_PARTS for part in rel.parts)


def file_map(root: Path) -> dict[str, str]:
    if not root.exists():
        return {}
    mapping: dict[str, str] = {}
    for file_path in sorted(root.rglob("*")):
        if not file_path.is_file() or _should_skip(file_path, root):
            continue
        rel = file_path.relative_to(root).as_posix()
        digest = hashlib.sha256(file_path.read_bytes()).hexdigest()
        mapping[rel] = digest
    return mapping


def hash_from_map(mapping: dict[str, str]) -> str:
    payload = json.dumps(mapping, sort_keys=True, separators=(",", ":"))
    digest = hashlib.sha256(payload.encode()).hexdigest()
    return f"sha256:{digest}"


def dir_hash(root: Path) -> str:
    return hash_from_map(file_map(root))


def remote_fingerprint(blob_map: dict[str, str]) -> str:
    return hash_from_map(blob_map)
